### PR TITLE
Created unhealthy  borrow reporter for Compound XY, Aave XY and Maker strategies

### PR DESCRIPTION
**Pull request by virendrapatidar**
- Added support for Compound (and fork) XY strategy for both mainnet and avalanche.
- Added support for Aave XY strategy for both mainnet and avalanche.
- Added support for Maker lowWater case for both mainnet and avalanche.
- Support pool Version 3, 4 and 5
- One agent covers all critical 

Sample Alert Logs tested at a block (by negating the final alert check).
```
 findings for block 0x8fddecfcda209c22957aedd96b8f23ae48aedbfc9a792ec1e367477f9e724ce1 {
  "name": "Strategy borrow more than allowed limit.",
  "description": "Strategy borrow more than allowed limit. network: Ethereum, strategyName: VesperAaveXYStrategyETH_FEI, strategyAddress: 0xa14d4c41b989dc4ae20150aa2ec7a6c6b5622406, poolName: vaETH Pool, poolAddress: 0xd1C117319B3595fbc39b471AB1fd485629eb05F2, ",
  "alertId": "Vesper",
  "protocol": "Vesper",
  "severity": "Critical",
  "type": "Info",
  "metadata": {
    "network": "Ethereum",
    "strategyName": "VesperAaveXYStrategyETH_FEI",
    "strategyAddress": "0xa14d4c41b989dc4ae20150aa2ec7a6c6b5622406",
    "poolName": "vaETH Pool",
    "poolAddress": "0xd1C117319B3595fbc39b471AB1fd485629eb05F2"
  },
  "addresses": []
},{
  "name": "Strategy borrow more than allowed limit.",
  "description": "Strategy borrow more than allowed limit. network: Ethereum, strategyName: VesperCompoundXYStrategyETH_LINK, strategyAddress: 0x7977581a346d330068fd5122f56fb1b4dc4e0ff9, poolName: vaETH Pool, poolAddress: 0xd1C117319B3595fbc39b471AB1fd485629eb05F2, ",
  "alertId": "Vesper",
  "protocol": "Vesper",
  "severity": "Critical",
  "type": "Info",
  "metadata": {
    "network": "Ethereum",
    "strategyName": "VesperCompoundXYStrategyETH_LINK",
    "strategyAddress": "0x7977581a346d330068fd5122f56fb1b4dc4e0ff9",
    "poolName": "vaETH Pool",
    "poolAddress": "0xd1C117319B3595fbc39b471AB1fd485629eb05F2"
  },
  "addresses": []
},{
  "name": "Strategy borrow more than allowed limit.",
  "description": "Strategy borrow more than allowed limit. network: Ethereum, strategyName: VesperCompoundXYStrategyWBTC_LINK, strategyAddress: 0x22d37d12ea1a3f18b8d4e2e3be844b6205590f1f, poolName: vaWBTC Pool, poolAddress: 0x01e1d41C1159b745298724c5Fd3eAfF3da1C6efD, ",
  "alertId": "Vesper",
  "protocol": "Vesper",
  "severity": "Critical",
  "type": "Info",
  "metadata": {
    "network": "Ethereum",
    "strategyName": "VesperCompoundXYStrategyWBTC_LINK",
    "strategyAddress": "0x22d37d12ea1a3f18b8d4e2e3be844b6205590f1f",
    "poolName": "vaWBTC Pool",
    "poolAddress": "0x01e1d41C1159b745298724c5Fd3eAfF3da1C6efD"
  },
  "addresses": []
},{
  "name": "Strategy borrow more than allowed limit.",
  "description": "Strategy borrow more than allowed limit. network: Ethereum, strategyName: VesperCompoundXYStrategyWBTC_USDC, strategyAddress: 0x0ee5f180868fb3a2c3a6d6b956d365879d826130, poolName: vaWBTC Pool, poolAddress: 0x01e1d41C1159b745298724c5Fd3eAfF3da1C6efD, ",
  "alertId": "Vesper",
  "protocol": "Vesper",
  "severity": "Critical",
  "type": "Info",
  "metadata": {
    "network": "Ethereum",
    "strategyName": "VesperCompoundXYStrategyWBTC_USDC",
    "strategyAddress": "0x0ee5f180868fb3a2c3a6d6b956d365879d826130",
    "poolName": "vaWBTC Pool",
    "poolAddress": "0x01e1d41C1159b745298724c5Fd3eAfF3da1C6efD"
  },
  "addresses": []
}

findings for block 0x8fddecfcda209c22957aedd96b8f23ae48aedbfc9a792ec1e367477f9e724ce1 {
  "name": "Strategy borrow more than allowed limit.",
  "description": "Strategy borrow more than allowed limit. network: Ethereum, strategyName: VesperMakerStrategyWBTC, strategyAddress: 0x70900c53202297bd7cd4ed1635486d84e3982484, poolName: vaWBTC Pool, poolAddress: 0x01e1d41C1159b745298724c5Fd3eAfF3da1C6efD, ",
  "alertId": "Vesper",
  "protocol": "Vesper",
  "severity": "Critical",
  "type": "Info",
  "metadata": {
    "network": "Ethereum",
    "strategyName": "VesperMakerStrategyWBTC",
    "strategyAddress": "0x70900c53202297bd7cd4ed1635486d84e3982484",
    "poolName": "vaWBTC Pool",
    "poolAddress": "0x01e1d41C1159b745298724c5Fd3eAfF3da1C6efD"
  },
  "addresses": []
}

```

closes https://github.com/bloqpriv/vesper-pools/issues/1357

### CODE REVIEW

File: Vesper-Agents-Suite/xy-safe-borrow-agent/src/agent.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  V3 = Array.from(new Set<string>(V3));
+  const valueV3promise = V3.map(async (strategy: string) => {
+    const strategyContract = new web3.eth.Contract(
+      STRATEGY_ABI,
+      strategy,
+    );
+
+    const name: string = await strategyContract.methods.NAME().call({}, blockNumber);
+    // Compound or it's forked XY Strategies 
+    if (!name.includes(XY_STRATEGY_NAME) || name.includes(Aave)) return BigInt(0);
```
Comment (patidarmanoj10) [inline Vesper-Agents-Suite/xy-safe-borrow-agent/src/agent.ts:43]: You need to handle AaveXY also. This if else will be true for strategy name VesperAaveXYStrategyETH_DAI also and next line will throw error.

Comment (patidarmanoj10) [inline Vesper-Agents-Suite/xy-safe-borrow-agent/src/agent.ts]:
```suggestion
    if (name.includes(XY_STRATEGY_NAME)) {
       const type = name.includes(Aave)? 'AaveXY' : 'CompoundLikeXY'
        return isBorrowBelowLimit(type)
    }
```

Comment (patidarmanoj10) [reply Vesper-Agents-Suite/xy-safe-borrow-agent/src/agent.ts]: Please add abstractions

File: Vesper-Agents-Suite/xy-safe-borrow-agent/src/agent.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    );
+
+    // get prices
+    const borrowTokenPrice = new BigNumber(await oracleContract.methods.getUnderlyingPrice(borrowCTokenAddress).call({}, blockNumber));
+    const supplyTokenPrice = new BigNumber(await oracleContract.methods.getUnderlyingPrice(supplyCTokenAddress).call({}, blockNumber));
+    const borrowed = new BigNumber(await borrowCToken.methods.borrowBalanceStored(strategy).call({}, blockNumber));
+    const supplyBalance = new BigNumber(await supplyCToken.methods.balanceOf(strategy).call({}, blockNumber));
+    const supplyTokenExchangeRate = new BigNumber(await supplyCToken.methods.exchangeRateStored().call({}, blockNumber));
+    const cf = (await comptrollerContract.methods.markets(supplyCTokenAddress).call({}, blockNumber)).collateralFactorMantissa
+    const maxBorrowLimit = new BigNumber(await strategyContract.methods.maxBorrowLimit().call({}, blockNumber));
```
Comment (patidarmanoj10) [inline Vesper-Agents-Suite/xy-safe-borrow-agent/src/agent.ts]: This is not good practice that you are using 10 await statement. Most of these can be fired simultaneously by using Promise.all()

File: Vesper-Agents-Suite/xy-safe-borrow-agent/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    const oracleAddress: string = await comptrollerContract.methods.oracle().call({}, blockNumber);
+    const oracleContract = new web3.eth.Contract(
+      ORACLE_ABI,
+      oracleAddress,
+    );
+
+    // get prices
+    const borrowTokenPrice = new BigNumber(await oracleContract.methods.getUnderlyingPrice(borrowCTokenAddress).call({}, blockNumber));
+    const supplyTokenPrice = new BigNumber(await oracleContract.methods.getUnderlyingPrice(supplyCTokenAddress).call({}, blockNumber));
+    const borrowed = new BigNumber(await borrowCToken.methods.borrowBalanceStored(strategy).call({}, blockNumber));
```
Comment (patidarmanoj10) [inline Vesper-Agents-Suite/xy-safe-borrow-agent/src/agent.ts]: why do you need to create BigNumber?  isnt borrowCToken.methods.borrowBalanceStored(strategy).call return bigNumber()?

**Comment by patidarmanoj10** [comment]
Rename agent name .  A noun name is preferred instead of verb .  High-borrow is verb .  higher-borrow-agent is noun.

**Comment by patidarmanoj10** [comment]
I don not have any suggestion but high-borrow is not fitting well.

**Comment by virendrapatidar** [comment]
> I don not have any suggestion but high-borrow is not fitting well.
The agent name is safe-borrow. Do not go with branch name. I discussed it with Rohit in slack and we agree on using safe borrow name.

**Review by patidarmanoj10 — APPROVED** [review]

**Pull request merged**